src/System.Net.Security/src/System/Net/Security/SNIHelper.cs

```diff
@@ -0,0 +1,251 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace System.Net.Security
+{
+    internal class SNIHelper
+    {
+        public static string GetServerName(byte[] clientHello)
+        {
+            return GetSniFromSslPlainText(clientHello);
+        }
+
+        // https://tools.ietf.org/html/rfc6101#section-5.2.1
+        // SSLPlainText structure:
+        //   - ContentType (1 byte) => 0x16 is handshake
+        //   - ProtocolVersion version (2 bytes)
+        //   - uint16 length
+        //   - opaque fragment[SSLPlaintext.length]
+        private static string GetSniFromSslPlainText(ReadOnlySpan<byte> sslPlainText)
+        {
+            // Is SSL 3 handshake? SSL 2 does not support extensions - skipping as well
+            if (sslPlainText.Length < 5 || sslPlainText[0] != 0x16)
+            {
+                return null;
+            }
+
+            ushort handshakeLength = ReadUint16(sslPlainText.Slice(3));
+            ReadOnlySpan<byte> sslHandshake = sslPlainText.Slice(5);
+
+            if (handshakeLength != sslHandshake.Length)
+            {
+                return null;
+            }
+
+            return GetSniFromSslHandshake(sslHandshake);
```

Comment:
Nit, might be simpler to just combine the methods

Reply:
Then you don't need to check `sslHandshake.Length < 4` I think.

Reply:
@danmosemsft, could you expand? I think I will still need to check - note that sslHandshake is sliced sslPlaintext. The lengths are fairly redundant in the SSL spec which makes this code slightly confusing but allow to skip parts implementation doesn't understand

Reply:
You are right. I misread.